ir/function.cpp

```diff
@@ -293,12 +362,129 @@ bool Function::removeUnusedStuff(const multimap<Value*, Value*> &users,
   return changed;
 }
 
+static string_view bb_dot_name(const string &name) {
+  if (name[0] == '%')
+    return string_view(name).substr(1);
+  return name;
+}
+
 void Function::unroll(unsigned k) {
   if (k == 0)
     return;
   LoopAnalysis la(*this);
   ofstream out("a.gv");
   la.printDot(out);
+
+  auto &Nodes = la.getHeaderNodes();
+  const auto &[roots, bottomup, topdown] = la.getNestedTree();
+
+  BasicBlock &sink = getBB("#sink");
+
+  auto Type = la.getNodeType();
+
+  vector<pair<const BasicBlock*, bool>> worklist;
+
+  for (auto &root : roots) {
+    worklist.emplace_back(root, false);
+  }
+
+  while (!worklist.empty()) {
+    auto &[bb, flag]  = worklist.back();
+    if (flag) {
+      worklist.pop_back();
+      map<Value*, Value*> vmap;
+      map<const BasicBlock *, vector<const BasicBlock *>> bbmap;
+      vector<const BasicBlock *> cloned;
+      auto bb_nodes = Nodes[bb];
+
+      for (auto &&bb_ii : bb_nodes) {
+        bbmap[bb_ii].push_back(bb_ii);
+      }
+      for (unsigned unroll = 1 ; unroll <= k; ++unroll) {
+        for (auto &&bb_ii : bb_nodes) {
+          auto &bb_i = const_cast<BasicBlock&>(*bb_ii);
+          string twine = "_" + string(bb_dot_name(bb->getName())) +
+                         "_" + to_string(unroll);
+          auto bb_cloned = &cloneBB(bb_i, twine, vmap);
+          bbmap[bb_ii].push_back(bb_cloned);
+          cloned.push_back(bb_cloned);
+        }
+      }
+
+      vector<const BasicBlock *> backedge_srcs;
+      set<const BasicBlock *> nodes_set(bb_nodes.begin(), bb_nodes.end());
+      // rewire targets
+      for (unsigned unroll = 0 ; unroll <= k; ++unroll) {
+        for (auto &&bb_ii : bb_nodes) {
+          auto curr = const_cast<BasicBlock *>(bbmap[bb_ii][unroll]);
+          map<const BasicBlock *, const BasicBlock *> replacement;
+          for (auto &tgt : curr->targets()) {
+            if (&tgt == bb || bb_ii != bb)
+              backedge_srcs.push_back(bb_ii);
+            // Filter out exit edges.
+            if (!nodes_set.count(&tgt))
+              continue;
+
+            const BasicBlock *to = nullptr;
+            // backedge
+            if (&tgt == bb) {
+              if (unroll == k)
+                to = &sink;
+              else
+                to = bbmap[&tgt][unroll + 1];
+            }
+            // edges inside loop
+            else {
+              to = bbmap[&tgt][unroll];
+            }
+            replacement[&tgt] = to;
+          }
+          for (auto &[from, to] : replacement)
+            curr->replaceTargetWith(from, to);
+        }
+      }
+
+      // rewire sources on header
+      auto bb_c = const_cast<BasicBlock*>(bb);
+
+      for (auto &src : backedge_srcs) {
+        bb_c->prunePhiSrc(src);
```

Comment:
What about adding `BasicBlock::removePhiSrcs`? It will be cheaper in total.

Reply:
The reason why it is cheaper is that, `removePhiSrc` iterates over the instructions in the basic block, and we don't want to repeat that for every item in `backedge_srcs`.

Reply:
If it leaves `removePhiSrcs` only and `removePhiSrc` is not used anymore, I think it is fine - all phi nodes should have well-formed incoming edges, anyway.